SQLite driver fails to compile on macOS: use of undeclared identifier 'sqlite3_load_extension'

The C SQLite driver's CMake build decides whether `sqlite3_load_extension()` is available by doing a plain-text search for that identifier in `sqlite3.h` (`c/driver/sqlite/CMakeLists.txt`).

That heuristic is unreliable: the string also appears in comments and inside `SQLITE_OMIT_LOAD_EXTENSION`-guarded blocks, so the symbol can be textually present in the header yet not actually declared. Apple's system SQLite is built with load-extension omitted, which fools the check into leaving `ADBC_SQLITE_WITH_NO_LOAD_EXTENSION` undefined, so `sqlite.cc` then fails to compile.

Found on this PR (with an unrelated change): https://github.com/apache/arrow-adbc/pull/4469, where [CI fails with](https://github.com/apache/arrow-adbc/actions/runs/28850435930/job/85565046218):

> [ 92%] Building C object driver/sqlite/CMakeFiles/adbc_driver_sqlite_objlib.dir/statement_reader.c.o
[ 92%] Building CXX object driver/sqlite/CMakeFiles/adbc_driver_sqlite_objlib.dir/sqlite.cc.o
/Users/runner/work/arrow-adbc/arrow-adbc/c/driver/sqlite/sqlite.cc:784:16: error: use of undeclared identifier 'sqlite3_load_extension'
784 | int rc = sqlite3_load_extension(conn_, extension_path_.c_str(),
| ^~~~~~~~~~~~~~~~~~~~~~
1 error generated.
make[2]: *** [driver/sqlite/CMakeFiles/adbc_driver_sqlite_objlib.dir/sqlite.cc.o] Error 1
make[1]: *** [driver/sqlite/CMakeFiles/adbc_driver_sqlite_objlib.dir/all] Error 2
make: *** [all] Error 2

Might be caused by a recent github CI runner image update: https://github.com/actions/runner-images/issues/14167

--- a/c/driver/sqlite/CMakeLists.txt
+++ b/c/driver/sqlite/CMakeLists.txt
@@ -30,14 +30,15 @@ else()
   set(SQLite3_INCLUDE_DIRS)
 endif()
 
-# Check for sqlite3_load_extension() in sqlite3.h
-if(EXISTS "${SQLite3_INCLUDE_DIRS}/sqlite3.h")
-  file(READ "${SQLite3_INCLUDE_DIRS}/sqlite3.h" ADBC_SQLITE_H_CONTENT)
-  string(FIND "${ADBC_SQLITE_H_CONTENT}" "sqlite3_load_extension"
-              ADBC_SQLITE_WITH_LOAD_EXTENSION)
-endif()
+# Check whether sqlite3_load_extension() is available.
+include(CheckSymbolExists)
+set(CMAKE_REQUIRED_INCLUDES ${SQLite3_INCLUDE_DIRS})
+set(CMAKE_REQUIRED_LIBRARIES ${SQLite3_LINK_LIBRARIES})
+check_symbol_exists(sqlite3_load_extension "sqlite3.h" ADBC_SQLITE_WITH_LOAD_EXTENSION)
+unset(CMAKE_REQUIRED_INCLUDES)
+unset(CMAKE_REQUIRED_LIBRARIES)
 
-if(ADBC_SQLITE_WITH_LOAD_EXTENSION EQUAL -1)
+if(NOT ADBC_SQLITE_WITH_LOAD_EXTENSION)
   set(ADBC_SQLITE_COMPILE_DEFINES "-DADBC_SQLITE_WITH_NO_LOAD_EXTENSION")
 endif()
 

--- a/c/driver/sqlite/CMakeLists.txt
+++ b/c/driver/sqlite/CMakeLists.txt
@@ -27,16 +27,26 @@ else()
   # vcpkg
   find_package(unofficial-sqlite3 CONFIG REQUIRED)
   set(SQLite3_LINK_LIBRARIES unofficial::sqlite3::sqlite3)
-  set(SQLite3_INCLUDE_DIRS)
+  get_target_property(SQLite3_INCLUDE_DIRS unofficial::sqlite3::sqlite3
+                      INTERFACE_INCLUDE_DIRECTORIES)
+  if(NOT SQLite3_INCLUDE_DIRS)
+    set(SQLite3_INCLUDE_DIRS)
+  endif()
 endif()
 
-# Check whether sqlite3_load_extension() is available.
+# Check whether sqlite3_load_extension() is declared in sqlite3.h. Compile a
+# static library instead of an executable so no linking happens: imported
+# targets in CMAKE_REQUIRED_LIBRARIES don't propagate into try_compile on
+# older CMake, and only the declaration matters (builds with
+# SQLITE_OMIT_LOAD_EXTENSION, like Apple's system SQLite, omit it).
 include(CheckSymbolExists)
+set(ADBC_SQLITE_SAVED_TRY_COMPILE_TARGET_TYPE ${CMAKE_TRY_COMPILE_TARGET_TYPE})
+set(CMAKE_TRY_COMPILE_TARGET_TYPE STATIC_LIBRARY)
 set(CMAKE_REQUIRED_INCLUDES ${SQLite3_INCLUDE_DIRS})
-set(CMAKE_REQUIRED_LIBRARIES ${SQLite3_LINK_LIBRARIES})
 check_symbol_exists(sqlite3_load_extension "sqlite3.h" ADBC_SQLITE_WITH_LOAD_EXTENSION)
 unset(CMAKE_REQUIRED_INCLUDES)
-unset(CMAKE_REQUIRED_LIBRARIES)
+set(CMAKE_TRY_COMPILE_TARGET_TYPE ${ADBC_SQLITE_SAVED_TRY_COMPILE_TARGET_TYPE})
+unset(ADBC_SQLITE_SAVED_TRY_COMPILE_TARGET_TYPE)
 
 if(NOT ADBC_SQLITE_WITH_LOAD_EXTENSION)
   set(ADBC_SQLITE_COMPILE_DEFINES "-DADBC_SQLITE_WITH_NO_LOAD_EXTENSION")

